compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,479 @@
+#[instrument(skip(tcx), level = "debug")]
+pub fn valtree_to_const_value<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env_ty: ty::ParamEnvAnd<'tcx, Ty<'tcx>>,
+    valtree: ty::ValTree<'tcx>,
+) -> ConstValue<'tcx> {
+    // Basic idea: We directly construct `Scalar` values from trivial `ValTree`s
+    // (those for constants with type bool, int, uint, float or char).
+    // For all other types we create an `MPlace` and fill that by walking
+    // the `ValTree` and using `place_projection` and `place_field` to
+    // create inner `MPlace`s which are filled recursively.
+    // FIXME Does this need an example?
+
+    let (param_env, ty) = param_env_ty.into_parts();
+    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
+
+    match ty.kind() {
+        ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) | ty::Char => match valtree {
+            ty::ValTree::Leaf(scalar_int) => ConstValue::Scalar(Scalar::Int(scalar_int)),
+            ty::ValTree::Branch(_) => bug!(
+                "ValTrees for Bool, Int, Uint, Float or Char should have the form ValTree::Leaf"
+            ),
+        },
+        ty::Ref(_, inner_ty, _) => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
+                    let slice_ty = match inner_ty.kind() {
+                        ty::Slice(slice_ty) => *slice_ty,
+                        ty::Str => tcx.mk_ty(ty::Uint(ty::UintTy::U8)),
+                        _ => bug!("expected ty::Slice | ty::Str"),
+                    };
+                    debug!(?slice_ty);
+
+                    let valtrees = valtree.unwrap_branch();
+
+                    // Create a place for the underlying array
+                    let len = valtrees.len();
+                    let arr_ty = tcx.mk_array(slice_ty, len as u64);
+                    let mut place =
+                        create_mplace_from_layout(&mut ecx, ty::ParamEnv::empty().and(arr_ty));
+                    debug!(?place);
+
+                    // Insert elements of `arr_valtree` into `place`
+                    fill_place_recursively(&mut ecx, &mut place, valtree, arr_ty);
+                    dump_place(&ecx, place.into());
+
+                    // The allocation behind `place` is local, we need to intern it
+                    intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &place).unwrap();
+
+                    // Now we need to get the Allocation
+                    let alloc_id = place.mplace.ptr.provenance.unwrap();
+                    debug!(?alloc_id);
+
+                    let data = match tcx.get_global_alloc(alloc_id) {
+                        Some(GlobalAlloc::Memory(const_alloc)) => const_alloc,
+                        _ => bug!("expected memory allocation"),
+                    };
+                    debug!(?data);
+
+                    return ConstValue::Slice { data, start: 0, end: len as usize };
+                }
+                _ => {
+                    match valtree {
```

Comment:
Except for the interning, these arms are exactly the same. Why don't we need interning in the second branch? Please deduplicate these two branches